app/src/org/commcare/dalvik/activities/TemplatePrinterActivity.java

```diff
@@ -0,0 +1,310 @@
+                    getIntent().getExtras(),
+                    this
+            ).execute();
+        } else {
+            showErrorDialog(Localization.get("template.invalid"));
+        }
+    }
+
+    /**
+     * Generate a unique name for this print job, using the name of the template file and the date
+     *
+     * @param templateFilename the path to the given template file
+     */
+    private void generateJobName(String templateFilename) {
+        String inputWithoutExtension = templateFilename.substring(0,
+                templateFilename.lastIndexOf('.'));
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
+        String dateString = sdf.format(new Date());
+        jobName = inputWithoutExtension + "_" + dateString;
+    }
+
+    /**
+     * Called when TemplatePrinterTask finishes, with a result code indicating what happened. If
+     * a .html file of the filled-out template has been created and saved successfully, proceeds
+     * with printing. Otherwise, displays the appropriate error message.
+     */
+    @Override
+    public void onFinished(int result, String problemString) {
+        switch(result) {
+            case TemplatePrinterTask.SUCCESS:
+                doHtmlPrint();
+                break;
+            case TemplatePrinterTask.IO_ERROR:
+                showErrorDialog(Localization.get("print.io.error"));
+                break;
+            case TemplatePrinterTask.VALIDATION_ERROR_MUSTACHE:
+                showErrorDialog(Localization.get("template.malformed.mustache", new String[]{problemString}));
+                break;
+            case TemplatePrinterTask.VALIDATION_ERROR_CHEVRON:
+                showErrorDialog(Localization.get("template.malformed.chevron", new String[]{problemString}));
+        }
+    }
+
+    /**
+     * Displays an error dialog with the specified message.
+     * Activity will quit upon exiting the dialog.
+     *
+     * @param message Error message
+     */
+    private void showErrorDialog(String message) {
+
+        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(this)
+                .setTitle(R.string.error_occured)
+                .setMessage(message)
+                .setCancelable(false)
+                .setPositiveButton(
+                        R.string.ok,
+                        new DialogInterface.OnClickListener() {
+
+                            @Override
+                            public void onClick(DialogInterface dialog, int which) {
+                                dialog.dismiss();
+                                finish();
+                            }
+                        }
+                );
+        dialogBuilder.show();
+    }
+
+    /**
+     * Shows an alert dialog about the status of the print job.
+     * Activity will quit upon exiting the dialog.
+     *
+     * @param msg String message that should be shown on the alert
+     */
+    private void showAlertDialog(String msg) {
```

Comment:
Should be moved into PrintDocumentAdapterWrapper class, where it is used, or into a util class if others might find this useful.